pkg/safe/state_test.go

```diff
@@ -145,3 +148,105 @@ func TestListFilter(t *testing.T) {
 
 	assert.Equal(t, 3, all.Len())
 }
+
+type testSpec struct{}
+
+func (testSpec) DeepCopy() testSpec { return testSpec{} }
+
+type testResource = typed.Resource[testSpec, testExtension]
+
+type testExtension struct{}
+
+func (testExtension) ResourceDefinition() meta.ResourceDefinitionSpec {
+	return meta.ResourceDefinitionSpec{
+		Type:             "testResource",
+		DefaultNamespace: "default",
+	}
+}
+
+const ns = "testns"
+
+// brokenState ignores the resource.Kind params and always returns string resources.
+type brokenState struct {
+	inmem.State
+}
+
+func (st *brokenState) List(context.Context, resource.Kind, ...state.ListOption) (resource.List, error) {
+	return resource.List{Items: []resource.Resource{conformance.NewStrResource(ns, "test", "test")}}, nil
+}
+
+func (st *brokenState) Get(context.Context, resource.Pointer, ...state.GetOption) (resource.Resource, error) {
+	return conformance.NewStrResource(ns, "test", "test"), nil
+}
+
+func (st *brokenState) Watch(_ context.Context, _ resource.Pointer, ch chan<- state.Event, _ ...state.WatchOption) error {
+	return st.publishWrong(ch)
+}
+
+func (st *brokenState) WatchKind(_ context.Context, _ resource.Kind, ch chan<- state.Event, _ ...state.WatchKindOption) error {
+	return st.publishWrong(ch)
+}
+
+func (*brokenState) publishWrong(ch chan<- state.Event) error {
+	go func() {
+		ch <- state.Event{
+			Type:     state.Bootstrapped,
+			Resource: conformance.NewStrResource(ns, "test", "test"),
+		}
+	}()
+
+	return nil
+}
+
+func TestTypeValidation(t *testing.T) {
+	s := &brokenState{}
+	resource := testResource{}
+	channel := make(chan safe.WrappedStateEvent[*testResource])
+
+	type testCase = func(t *testing.T) (result any, err error)
+
+	cases := map[string]testCase{
+		"StateGet": func(_ *testing.T) (any, error) {
```

Comment:
```suggestion
		"StateGet": func(*testing.T) (any, error) {
```

Here and below.

Reply:
Ah, good point! addressed in [79ca968](https://github.com/cosi-project/runtime/pull/574/commits/79ca968c4c93d77238a19a57189a4eb849b4f3fd)